### PR TITLE
Refactor command utilities for tests

### PR DESCRIPTION
**Pull request by yjbanov**
## Description

Refactor command utilities for tests:

- Split `runCommand` into `startCommand` and `runCommand`. (`startCommand` will be used in https://github.com/flutter/flutter/pull/67324).
- Introduce:
  - `Command` represents a running command.
  - `CommandResult` represents the result of a finished command.
- Remove unused parameters from `runAndGetStdout` and reimplement it in terms of `runCommand`.
- Remove `CapturedOutput`. Reimplement existing usages in terms of `CommandResult`.
- Remove the no longer necessary `OutputMode.discard` (you "discard" by ignoring the `CommandResult`).

## Tests

Refactoring existing tests.

### CODE REVIEW

**Comment by yjbanov** [comment]
Sorry for back'n'forth. A last-minute "clean-up" ended up clashing with `dart:io` names. This should be good now.

File: dev/bots/run_command.dart [hunk 1]
```diff
@@ -2,92 +2,112 @@
...
+    outputMode: OutputMode.capture,
+    outputListener: (String line, io.Process process) {
+      output.add(line);
+    },
   );
 
-  stderr.addStream(process.stderr);
-  final Stream<String> lines = process.stdout.transform(utf8.decoder).transform(const LineSplitter());
-  yield* lines;
+  yield* output.stream;
```
Comment (jonahwilliams) [inline dev/bots/run_command.dart:40]: What does this actually do?

Comment (jonahwilliams) [reply dev/bots/run_command.dart:40]: I'm trying to mentally model this - we run the stream until it gets closed, then await the result?

But when does the stream get closed?

Comment (yjbanov) [reply dev/bots/run_command.dart:40]: I imagine the stream gets closed when the process exit. I just don't know which one resolves first the stream or the exit code :man_shrugging:

Comment (jonahwilliams) [reply dev/bots/run_command.dart:40]: But you created the output stream controller above, right?

Comment (yjbanov) [reply dev/bots/run_command.dart:40]: Are you implying that I need to explicitly close the controller? Hmm :thinking:

Comment (jonahwilliams) [reply dev/bots/run_command.dart:40]: other wise how does the stream end 😓

Comment (yjbanov) [reply dev/bots/run_command.dart:40]: Yep, you were right. Done.

File: dev/bots/run_command.dart [hunk 2]
```diff
@@ -2,92 +2,115 @@
...
   );
 
-  stderr.addStream(process.stderr);
-  final Stream<String> lines = process.stdout.transform(utf8.decoder).transform(const LineSplitter());
-  yield* lines;
+  // Close the stream controller after the command is complete. Otherwise,
+  // the yield* will never finish.
+  command.whenComplete(() {
+    output.close();
+  });
```
Comment (jonahwilliams) [inline dev/bots/run_command.dart:44]:
```suggestion
  command.whenComplete(output.close);
```

File: dev/bots/run_command.dart (hunk 1 above)
Comment (yjbanov) [reply dev/bots/run_command.dart]: For some reason I thought GC finalizes it... although now I realize that wouldn't work in JS until very recently.

**Review by jonahwilliams — APPROVED** [review]
LGTM if LG to CI

**Pull request merged**